scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -157,6 +167,70 @@ impl ReplicatorConsumer {
         Ok(())
     }
 
+    // Recreates INSERT/DELETE/UPDATE statement on non-frozen list.
+    async fn update_list<'a>(
+        &self,
+        column_name: &str,
+        data: &'a CDCRow<'_>,
+        values_for_update: &mut Vec<&'a CqlValue>,
+        values_for_remove: &mut Vec<&'a CqlValue>,
+        values_for_delete: &[&CqlValue],
+        timestamp: i64,
+    ) -> anyhow::Result<()> {
+        if data.is_value_deleted(column_name) {
+            // If the list was replaced with syntax "(...) SET l = [(...)]",
+            // we have to manually delete the list
+            // and add every element with timeuuid from the original table,
+            // because we want to preserve the timestamps of the list elements.
+            // The list is deleted with timestamp decreased by one,
+            // this is how Scylla does it internally.
+            self.run_statement(
+                Query::new(format!(
+                    "DELETE {col} FROM {ks}.{tbl} WHERE {cond}",
+                    col = column_name,
+                    ks = self.dest_keyspace_name,
+                    tbl = self.dest_table_name,
+                    cond = self.keys_cond,
+                )),
+                values_for_delete,
+                timestamp - 1,
+            )
+            .await?;
+        }
+
+        let add_query = Query::new(
+            format!("UPDATE {ks}.{tbl} USING TTL ? SET {list}[SCYLLA_TIMEUUID_LIST_INDEX(?)] = ? WHERE {cond}",
+                             ks = self.dest_keyspace_name,
+                             tbl = self.dest_table_name,
+                             list = column_name,
+                             cond = self.keys_cond));
+
+        if let Some(Map(added_elements)) = data.get_value(column_name) {
+            for (key, val) in added_elements {
+                values_for_update[1] = key;
+                values_for_update[2] = val;
+
+                self.run_statement(add_query.clone(), values_for_update, timestamp)
+                    .await?;
+            }
+        }
+        let remove_query = Query::new(
+            format!("UPDATE {ks}.{tbl} USING TTL ? SET {list}[SCYLLA_TIMEUUID_LIST_INDEX(?)] = null WHERE {cond}",
+                             ks = self.dest_keyspace_name,
+                             tbl = self.dest_table_name,
+                             list = column_name,
+                             cond = self.keys_cond));
```

Comment:
I tried to resolve these two actions by using one query (the `add_query`), but I couldn't pass `null` as `CqlValue` - when I set the value of `SET l[(...)] = ?` to `CqlValue::Empty`, I get the following error:
```
Database returned an error: Invalid protocol message received from the driver, Error message: truncated frame: expected 589829 bytes, length is 6
```
This is why there are two separate queries and two separate lists of arguments.

Reply:
`CqlValue::Empty` is not the same as `null`. To represent a nullable value, you can use `Option<CqlValue>`.

Reply:
Wouldn't I have to replace the type of the vector to `Vec<Option<CqlValue>>` then? Seems like trouble, not sure which version is better.
Btw. if `CqlValue::Empty` is not used for representing `null`, what is its purpose? I couldn't find anything in the documentation.

Reply:
`CqlValue::Empty` is a weird case in the Cassandra's/Scylla's type system. Most types, even if they are not conventionally a type that can be empty, can be set to an empty value. For example: `int`. An empty `int` is different from `null` and from `0`.

Reply:
I don't think that changing `values_for_update` to `Vec<Option<CqlValue>>` would be troublesome here, would it? You would also have to change `values_for_list_update`, but it is only used for lists so this won't affect other functions. If you do it, you will also be able to drop the `values_for_remove` argument.